apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryDataReader.cs

```diff
@@ -14,197 +14,207 @@
 
 using System;
 using System.Collections;
-using System.Data;
+using System.Collections.Generic;
 using System.Data.Common;
+using System.Threading;
+using System.Threading.Tasks;
+using Google.Api.Gax;
 
 namespace Google.Cloud.BigQuery.V2
 {
     /// <summary>
+    /// Reads a forward-only stream of rows from a data source.
     /// </summary>
     public sealed class BigQueryDataReader : DbDataReader
     {
-#if NET45
+        private readonly BigQueryConnection _connectionToClose;
+        private IAsyncEnumerator<BigQueryRow> _results;
+        private bool _isEnumerating;
 
+        internal BigQueryDataReader(BigQueryResults bigQueryResults, BigQueryConnection connectionToClose = null)
+        {
+            _connectionToClose = connectionToClose;
+            _isEnumerating = true;
+
+            GaxPreconditions.CheckNotNull(bigQueryResults, nameof(bigQueryResults));
+            BigQueryResults = bigQueryResults;
+        }
+
+
+#if NET45 
         /// <inheritdoc />
         public override void Close()
         {
-            throw new NotImplementedException();
+            _results?.Dispose();
+            _connectionToClose?.Close();
         }
 
         /// <inheritdoc />
-        public override DataTable GetSchemaTable()
+        public override System.Data.DataTable GetSchemaTable()
         {
-            throw new NotImplementedException();
+            // We cannot throw here as some scenarios (datatable) will call this method
+            return null;
         }
 #endif
 
         /// <inheritdoc />
-        public override bool NextResult()
-        {
-            throw new NotImplementedException();
-        }
+        public override bool NextResult() => false;
 
         /// <inheritdoc />
         public override bool Read()
-        {
-            throw new NotImplementedException();
-        }
+            => ReadAsync(CancellationToken.None).ResultWithUnwrappedExceptions();
 
         /// <inheritdoc />
-        public override int Depth { get; }
+        public override async Task<bool> ReadAsync(CancellationToken cancellationToken)
+            => _isEnumerating = await MoveNextAsync().ConfigureAwait(false);
+
         /// <inheritdoc />
-        public override bool IsClosed { get; }
+        public override int Depth => 0;
+
         /// <inheritdoc />
-        public override int RecordsAffected { get; }
+        public override bool HasRows => _isEnumerating && _results?.Current != null;
```

Comment:
So if we haven't called Read, this will return false always, because Current will return null - is that appropriate?

Reply:
hmm we have a choice of returning true or false before we actually know for sure what the value should be (when ReadAsync is called).  prefilling the state adds unnecessary complication and possible perf side effects -- so I don't want to do that.
I agree, returning "true" here is probably better than "false" -- but developers will need to handle the case of us returning True for HasRows and then immediately returning false for ReadAsync.  Not entirely clear, but I suppose better than the alternative.